### ISSUE TITLE
Deployments create too many pods

### ISSUE THREAD

**Issue by seanmck** [Azure/aci-connector-k8s#24]
When pods are created as replicas in a deployment, you end up with about 3x the number of replicas requested. Kubernetes starts by creating the right number of replicas, but as they reach the running state, new ones are unnecessarily created.

**Comment by seglauf** [comment]
Have the same Problem. 
replicas 2 -> stopped at 20... where 2 of them where allways pending...

**Comment by lachie83** [comment]
Thanks for reporting this. There must be some race condition in the logic. We'll investigate.

**Comment by lachie83** [comment]
I'm able to replicate this issue.

```
Events:
  FirstSeen	LastSeen	Count	From			SubObjectPath	Type		Reason			Message
  ---------	--------	-----	----			-------------	--------	------			-------
  4m		4m		1	replicaset-controller			Normal		SuccessfulCreate	Created pod: nuanced-sasquatch-minecraft-2061528660-gbmfw
  1m		1m		1	replicaset-controller			Normal		SuccessfulCreate	Created pod: nuanced-sasquatch-minecraft-2061528660-4xdgt
  1m		1m		1	replicaset-controller			Normal		SuccessfulCreate	Created pod: nuanced-sasquatch-minecraft-2061528660-wpcgs
  1m		1m		1	replicaset-controller			Normal		SuccessfulCreate	Created pod: nuanced-sasquatch-minecraft-2061528660-s8phq
  1m		1m		1	replicaset-controller			Normal		SuccessfulCreate	Created pod: nuanced-sasquatch-minecraft-2061528660-6w4xx
  1m		1m		1	replicaset-controller			Normal		SuccessfulCreate	Created pod: nuanced-sasquatch-minecraft-2061528660-j4km0
  1m		1m		1	replicaset-controller			Normal		SuccessfulCreate	Created pod: nuanced-sasquatch-minecraft-2061528660-j16w9
  59s		59s		1	replicaset-controller			Normal		SuccessfulCreate	Created pod: nuanced-sasquatch-minecraft-2061528660-17jt1
  44s		44s		1	replicaset-controller			Normal		SuccessfulCreate	Created pod: nuanced-sasquatch-minecraft-2061528660-jmwkw
  28s		8s		2	replicaset-controller			Normal		SuccessfulCreate	(combined from similar events): Created pod: nuanced-sasquatch-minecraft-2061528660-0bq20
```

I believe that the pod update message may not be annotated correctly so that the replicaset can detect it.

**Comment by JockDaRock** [comment]
I am also having this issue as well.  when I create a deployment a huge amount of containers are being spun up in the aci-connector node... when I set replicas to zero i still get a lot of pods but no pending containers.  If I set a number of replicas, whatever number that is, I will get that many pending, and then a bunch of pods to go with it.

**Comment by neilpeterson** [comment]
I am also seeing what I am assuming is this issue. Not sure if this is helpful, but the behavior in my environment is predictable / reproducible. Also to note, I've reproduced both with and without the canary build.

## Observations:

In my tests, pods are created unscheduled and then scheduled via. a custom scheduler / POST event.

- Create a deployment with an unscheduled pod - (pod 1).
- At the time that pod 1 is scheduled on the aci-connector node, pod 2 is 'auto-created' (unassigned).
- At this time, it looks like pod 1 (running) is removed from the replica set, and pod 2 (not-scheduled) is added to the replica set.
- It also appears that labels are removed from pod 1.
- If the deployment is deleted, pod 2 is deleted, but pod 1 is not (assuming related to being removed from the replica set).
- If I were to scheduled pod 2, the issue 'loops', creating a third pod (not scheduled) joined to the replica set. Pod 1 and 2 are both running and have been removed from the replica set. When the deployment is deleted, only pod 3 is deleted.

Pod 1 - pre-scheduled:

```
{
  "kind": "Pod",
  "apiVersion": "v1",
  "metadata": {
    "name": "azure-vote-front-614820086-dlkxx",
    "generateName": "azure-vote-front-614820086-",
    "namespace": "default",
    "selfLink": "/api/v1/namespaces/default/pods/azure-vote-front-614820086-dlkxx",
    "uid": "30733ab0-92b7-11e7-b8bf-000d3a16bcaf",
    "resourceVersion": "45619",
    "creationTimestamp": "2017-09-06T03:55:14Z",
    "labels": {
      "app": "azure-vote-front",
      "pod-template-hash": "614820086"
    },
    "annotations": {
      "kubernetes.io/created-by": "{\"kind\":\"SerializedReference\",\"apiVersion\":\"v1\",\"reference\":{\"kind\":\"ReplicaSet\",\"namespace\":\"default\",\"name\":\"azure-vote-front-614820086\",\"uid\":\"3071a773-92b7-11e7-b8bf-000d3a16bcaf\",\"apiVersion\":\"extensions\",\"resourceVersion\":\"45618\"}}\n"
    },
    "ownerReferences": [
      {
        "apiVersion": "extensions/v1beta1",
        "kind": "ReplicaSet",
        "name": "azure-vote-front-614820086",
        "uid": "3071a773-92b7-11e7-b8bf-000d3a16bcaf",
        "controller": true,
        "blockOwnerDeletion": true
      }
    ]
  },
  "spec": {
    "volumes": [
      {
        "name": "default-token-m4qg7",
        "secret": {
          "secretName": "default-token-m4qg7",
          "defaultMode": 420
        }
      }
    ],
    "containers": [
      {
        "name": "azure-vote-front",
        "image": "microsoft/azure-vote-front:redis-v1",
        "ports": [
          {
            "containerPort": 80,
            "protocol": "TCP"
          }
        ],
        "env": [
          {
            "name": "REDIS",
            "value": "azure-vote-back"
          }
        ],
        "resources": {},
        "volumeMounts": [
          {
            "name": "default-token-m4qg7",
            "readOnly": true,
            "mountPath": "/var/run/secrets/kubernetes.io/serviceaccount"
          }
        ],
        "terminationMessagePath": "/dev/termination-log",
        "terminationMessagePolicy": "File",
        "imagePullPolicy": "IfNotPresent"
      }
    ],
    "restartPolicy": "Always",
    "terminationGracePeriodSeconds": 30,
    "dnsPolicy": "ClusterFirst",
    "serviceAccountName": "default",
    "serviceAccount": "default",
    "securityContext": {},
    "schedulerName": "aci-burst"
  },
  "status": {
    "phase": "Pending",
    "qosClass": "BestEffort"
  }
}
```

Pod 1 - post-scheduled (running):

```
{
  "kind": "Pod",
  "apiVersion": "v1",
  "metadata": {
    "name": "azure-vote-front-614820086-dlkxx",
    "namespace": "default",
    "selfLink": "/api/v1/namespaces/default/pods/azure-vote-front-614820086-dlkxx",
    "uid": "30733ab0-92b7-11e7-b8bf-000d3a16bcaf",
    "resourceVersion": "46121",
    "creationTimestamp": "2017-09-06T03:55:14Z"
  },
  "spec": {
    "volumes": [
      {
        "name": "default-token-m4qg7",
        "secret": {
          "secretName": "default-token-m4qg7",
          "defaultMode": 420
        }
      }
    ],
    "containers": [
      {
        "name": "azure-vote-front",
        "image": "microsoft/azure-vote-front:redis-v1",
        "ports": [
          {
            "containerPort": 80,
            "protocol": "TCP"
          }
        ],
        "env": [
          {
            "name": "REDIS",
            "value": "azure-vote-back"
          }
        ],
        "resources": {},
        "volumeMounts": [
          {
            "name": "default-token-m4qg7",
            "readOnly": true,
            "mountPath": "/var/run/secrets/kubernetes.io/serviceaccount"
          }
        ],
        "terminationMessagePath": "/dev/termination-log",
        "terminationMessagePolicy": "File",
        "imagePullPolicy": "IfNotPresent"
      }
    ],
    "restartPolicy": "Always",
    "terminationGracePeriodSeconds": 30,
    "dnsPolicy": "ClusterFirst",
    "serviceAccountName": "default",
    "serviceAccount": "default",
    "nodeName": "aci-connector",
    "securityContext": {},
    "schedulerName": "aci-burst"
  },
  "status": {
    "phase": "Running",
    "conditions": [
      {
        "type": "Initialized",
        "status": "True",
        "lastProbeTime": null,
        "lastTransitionTime": "2017-09-06T03:39:27Z"
      },
      {
        "type": "PodScheduled",
        "status": "True",
        "lastProbeTime": null,
        "lastTransitionTime": "2017-09-06T03:39:27Z"
      },
      {
        "type": "Ready",
        "status": "True",
        "lastProbeTime": null,
        "lastTransitionTime": "2017-09-06T03:39:27Z"
      }
    ],
    "podIP": "40.85.155.4",
    "startTime": "2017-09-06T03:39:27Z",
    "containerStatuses": [
      {
        "name": "azure-vote-front",
        "state": {
          "running": {
            "startedAt": "2017-09-06T03:39:27Z"
          }
        },
        "lastState": {},
        "ready": true,
        "restartCount": 0,
        "image": "microsoft/azure-vote-front:redis-v1",
        "imageID": ""
      }
    ]
  }
}
```

Pod 2 - never scheduled:

```
{
  "kind": "Pod",
  "apiVersion": "v1",
  "metadata": {
    "name": "azure-vote-front-614820086-kzz5t",
    "generateName": "azure-vote-front-614820086-",
    "namespace": "default",
    "selfLink": "/api/v1/namespaces/default/pods/azure-vote-front-614820086-kzz5t",
    "uid": "d684708c-92b7-11e7-b8bf-000d3a16bcaf",
    "resourceVersion": "46122",
    "creationTimestamp": "2017-09-06T03:59:53Z",
    "labels": {
      "app": "azure-vote-front",
      "pod-template-hash": "614820086"
    },
    "annotations": {
      "kubernetes.io/created-by": "{\"kind\":\"SerializedReference\",\"apiVersion\":\"v1\",\"reference\":{\"kind\":\"ReplicaSet\",\"namespace\":\"default\",\"name\":\"azure-vote-front-614820086\",\"uid\":\"3071a773-92b7-11e7-b8bf-000d3a16bcaf\",\"apiVersion\":\"extensions\",\"resourceVersion\":\"45624\"}}\n"
    },
    "ownerReferences": [
      {
        "apiVersion": "extensions/v1beta1",
        "kind": "ReplicaSet",
        "name": "azure-vote-front-614820086",
        "uid": "3071a773-92b7-11e7-b8bf-000d3a16bcaf",
        "controller": true,
        "blockOwnerDeletion": true
      }
    ]
  },
  "spec": {
    "volumes": [
      {
        "name": "default-token-m4qg7",
        "secret": {
          "secretName": "default-token-m4qg7",
          "defaultMode": 420
        }
      }
    ],
    "containers": [
      {
        "name": "azure-vote-front",
        "image": "microsoft/azure-vote-front:redis-v1",
        "ports": [
          {
            "containerPort": 80,
            "protocol": "TCP"
          }
        ],
        "env": [
          {
            "name": "REDIS",
            "value": "azure-vote-back"
          }
        ],
        "resources": {},
        "volumeMounts": [
          {
            "name": "default-token-m4qg7",
            "readOnly": true,
            "mountPath": "/var/run/secrets/kubernetes.io/serviceaccount"
          }
        ],
        "terminationMessagePath": "/dev/termination-log",
        "terminationMessagePolicy": "File",
        "imagePullPolicy": "IfNotPresent"
      }
    ],
    "restartPolicy": "Always",
    "terminationGracePeriodSeconds": 30,
    "dnsPolicy": "ClusterFirst",
    "serviceAccountName": "default",
    "serviceAccount": "default",
    "securityContext": {},
    "schedulerName": "aci-burst"
  },
  "status": {
    "phase": "Pending",
    "qosClass": "BestEffort"
  }
}
```

**Comment by lachie83** [comment]
Fixed in #54
